### PR TITLE
POLLRDHUP delivery fix

### PR DESCRIPTION
**Pull request by turekt**
Hi,

I am proposing a change that should fix issue #6015. 

As per my understanding POLLRDHUP triggers when there is a half-closed TCP connection (either remote shutdown(SHUT\_WR) or local shutdown(SHUT\_RD)), therefore I have added a new equivalent event to waiter - `waiter.EventRdHUp` and added the event to `allEvents` and a "list" of events that can get triggered on read in `sys_read.go`. The TCP endpoint now checks whether the connection is half-closed and sets the appropriate flag. Additionally, TCP shutdown function notifies when only read end of the endpoint connection is closed. I have updated the syscall tests to reflect this change.

It is possible that some things are missing here but I will gladly update the PR if something was overlooked.

Let me know what you think.

### CODE REVIEW

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (hbhasker) [inline pkg/tcpip/transport/tcp/endpoint.go:957]: Readiness cannot acquire e.mu. This leads to a deadlock in epoll. I would suggest creating a flags field in the endpoint that is read atomically and use that instead of RcvClosed/SndClosed to identify when to deliver RdHup.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (hbhasker) [inline pkg/tcpip/transport/tcp/endpoint.go:2577]: I am not sure this is correct. Shutting down write should not result in a RdHUP as that is meant to be delivered when we receive a FIN from the peer to indicate no more data is expected from the peer.

Comment (hbhasker) [reply pkg/tcpip/transport/tcp/endpoint.go:2577]: You probably want to notify this in rcv.handleSegments where we process the incoming FIN.

**Review by hbhasker — COMMENTED** [review]
Thanks for the fix. There are a couple of changes that are needed before this can be merged.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (turekt) [reply pkg/tcpip/transport/tcp/endpoint.go]: Good point. I missed the endpoint struct Godoc that mentioned exactly that. Done.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (turekt) [reply pkg/tcpip/transport/tcp/endpoint.go:2589]: I am not sure if we are on the same page here. The newly introduced code checks whether the `tcpip.ShutdownWrite` is **not** set in the `shutdownFlags` at a point where `shutdownFlags` most certainly have `tcpip.ShutdownRead` flag set (see line 2556). This would mean that the `shutdownLocked` function will close the endpoint for read but write will stay open resulting in a half-closed TCP connection with write opened and read closed, which should trigger POLLRDHUP by definition. Linux implementation of TCP uses a similar approach by [checking shutdown mask to set POLLRDHUP flag](https://github.com/torvalds/linux/blob/f348d70a324e15afc701a494f32ec468abb7d1eb/net/ipv4/tcp.c#L367-L368).

Nevertheless, it is possible that I am overthinking everything and if you would still like me to change this to a different approach, I would be happy to do it.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (hbhasker) [inline pkg/tcpip/transport/tcp/endpoint.go:3246]: I believe your PR needs to be rebased @ HEAD. This code has changed. Further I don't think we care about RdHup here as we do want the endpoint to be fully closed when Wait returns.

**Review by hbhasker — COMMENTED** [review]
Let's add another test to socket_inet_loopback.cc or tcp_socket.cc that verifies the following for a pair of connected sockets.

a) A thread waiting on socket 1 w/ epoll for POLLIN
b) call shutdown(sock1, SHUTDOWN_RD)
c) a) should unblock due to POLLRDHUP.

Another test
a) a thread waiting on POLLIN on sock1
b) socket 2 does an active close or shutdown(sock2, SHUTDOWN_WR)
c) sock1 should wake up.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (turekt) [reply pkg/tcpip/transport/tcp/endpoint.go]: Removed and rebased.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (turekt) [reply pkg/tcpip/transport/tcp/endpoint.go]: Done.

**Comment by turekt** [comment]
I have added the tests. Let me know what you think.

**Comment by turekt** [comment]
Hi @hbhasker, I understand that you might be busy. I have noticed that it is soon going to be a year since this PR was created. I was wondering whether you could find the time to review this PR?

Thank you in advance.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (kevinGC) [inline pkg/tcpip/transport/tcp/endpoint.go:80]: Can we type these and add a named zero-value?

```
type connDirectionState int

// Connection direction states used for directionState checks in endpoint struct
// to detect half-closed connection and deliver POLLRDHUP
const (
	connDirectionStateOpen connDirectionState = 0
	// ...
```

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (kevinGC) [inline pkg/tcpip/transport/tcp/endpoint.go:950]: These can be one line: `if cs := e.connDirectionState(); cs&connDirectionStateAll == connDirectionStateRcvClosed {`

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (kevinGC) [inline pkg/tcpip/transport/tcp/endpoint.go:2624]: `connDirectionStateOpen` instead of 0.

**Review by kevinGC — DISMISSED** [review]
I'm taking over as reviewer.

Looks good, mostly test changes. We've had flakiness problems in the past, so we try to be extra careful about timeouts/sleeps/etc.

File: test/syscalls/linux/socket_inet_loopback.cc
Comment (kevinGC) [reply test/syscalls/linux/socket_inet_loopback.cc:458]: Note: this applies to the 2nd test ask well.

File: test/syscalls/linux/socket_inet_loopback.cc
Comment (turekt) [reply test/syscalls/linux/socket_inet_loopback.cc]: Good point. Refactored.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (turekt) [reply pkg/tcpip/transport/tcp/endpoint.go:82]: Done.

**Comment by turekt** [comment]
Hi @kevinGC, thanks for your review. I have changed the code per your suggestions.

Let me know what you think.

**Review by kevinGC — COMMENTED** [review]
LGTM

**Comment by kevinGC** [comment]
Been trying to merge this, but running into a failing test internally. Should be able to reproduce it by running this:

```
bazel test //test/syscalls:socket_inet_loopback_test_runsc_ptrace --runs_per_test=1000
```

`POLLRDHUP` isn't being returned all the time -- it's flaky.

I believe this is happening because the endpoint state `e.state` changes independently of `e.connectionDirectionState`. Because they aren't synchronized, it's possible for this sequence to occur:

- `e.state` is updated
- `e.Readiness` is called, and returns only `POLLIN`/`EventIn`
- The test `poll()` returns only `POLLIN`
- `e.connectionDirectionState` is set to `connDirectionStateRcvClosed`, but it's too late

We might have to synchronize these values, or take a look at how others handle the problem. Not 100% sure this is the issue, but it's my current theory.

**Comment by turekt** [comment]
Hi @kevinGC, thanks for letting me know!

I managed to reproduce the failing test. I made a small change to the commit and retested. My first test run finished successfully, but I am running a second one just in case. The `--runs_per_test=1000` is very slow for me so I was hoping that we could quickly verify the second test run with copybara if possible.

Hopefully this change fixes the POLLRDHUP issue.

**Comment by turekt** [comment]
Hi @kevinGC, the second test run is finished and successful! The latest changes should be a complete fix for the POLLRDHUP issue.

**Review by kevinGC — COMMENTED** [review]
I'm sorry for the very late review.

File: pkg/tcpip/transport/tcp/endpoint.go
Comment (turekt) [reply pkg/tcpip/transport/tcp/endpoint.go]: Good catch! Changed.

**Comment by turekt** [comment]
Hi @kevinGC, any updates?

**Pull request merged**
